From 6942bb814dcf9361dc9c3dac395842444159be54 Mon Sep 17 00:00:00 2001
From: Georges Basile Stavracas Neto <georges.stavracas@gmail.com>
Date: Tue, 22 Jun 2021 22:31:26 -0300
Subject: [PATCH] pipewire: Properly account for cursor hotspot

The cursor bitmap is centered on the hotspot, so not accounting
for it means PipeWire captures were positioning the cursor sprite
slightly off.

Properly account for the hotspot by subtracting it from the cursor
position.

Related: https://github.com/obsproject/obs-studio/issues/4766
---
 plugins/linux-capture/pipewire.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/plugins/linux-capture/pipewire.c b/plugins/linux-capture/pipewire.c
index abd0d4923..114e66c33 100644
--- a/plugins/linux-capture/pipewire.c
+++ b/plugins/linux-capture/pipewire.c
@@ -1234,9 +1234,11 @@ void obs_pipewire_video_render(obs_pipewire_data *obs_pw, gs_effect_t *effect)
 
 	if (obs_pw->cursor.visible && obs_pw->cursor.valid &&
 	    obs_pw->cursor.texture) {
+		float cursor_x = obs_pw->cursor.x - obs_pw->cursor.hotspot_x;
+		float cursor_y = obs_pw->cursor.y - obs_pw->cursor.hotspot_y;
+
 		gs_matrix_push();
-		gs_matrix_translate3f((float)obs_pw->cursor.x,
-				      (float)obs_pw->cursor.y, 0.0f);
+		gs_matrix_translate3f(cursor_x, cursor_y, 0.0f);
 
 		gs_effect_set_texture(image, obs_pw->cursor.texture);
 		gs_draw_sprite(obs_pw->texture, 0, obs_pw->cursor.width,
-- 
2.31.1

